### PR TITLE
Oop design principles

### PR DESCRIPTION
**Pull request by liubamiron**

### CODE REVIEW

File: app/Services/ModelLogger.php [hunk 1]
```diff
@@ -31,6 +31,17 @@ public function logModel(?User $user, LoggableInterface $loggable): void
                  );
             }
 
+            public function logModel2(?User $user, LoggableInterface $loggable): void
```
Comment (egorbwork) [inline app/Services/ModelLogger.php]: Not acceptable

File: app/Services/ProductionRequestActivityLogger.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace App\Services;
+
+use Illuminate\Http\Request;
+
+class ProductionRequestActivityLogger extends AbstractRequestActivityLogger
+{
+    protected function collectRequestData(Request $request): array
+    {
+        return [$request->email,
```
Comment (egorbwork) [inline app/Services/ProductionRequestActivityLogger.php]: request data should not be collected in the Production logger

File: views/imobils/test.blade.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
+         
+          <div class="card shadow-sm">
```
Comment (egorbwork) [inline views/imobils/test.blade.php]: Should not be part of the pr

File: app/Services/DebugRequestActivityLogger.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace App\Services;
+
+use Illuminate\Http\Request;
+
+class DebugRequestActivityLogger extends AbstractRequestActivityLogger
+{
+    protected function collectRequestData(Request $request): array
+    {
+        return [$request->email,
```
Comment (egorbwork) [inline app/Services/DebugRequestActivityLogger.php]: For debug all data can be collected in a single key, also other information should be collected.

File: app/Http/Providers/AppServiceProvider.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public function register()
+    {
+        //  $this->app->bind(RequestActivityLoggerInterface:: class, function(){
+         // return $this->app->make(DummyRequestActivityLogger::class);
+
+        
+
+
+      $this->app->bind(RequestActivityLoggerInterface::class, DummyRequestActivityLogger::class);
```
Comment (egorbwork) [inline app/Http/Providers/AppServiceProvider.php]: keeping both definition is mistake, this one should be removed.

File: app/Http/Providers/AppServiceProvider.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    if (config('app.env')==='production')  {
```
Comment (egorbwork) [inline app/Http/Providers/AppServiceProvider.php]: I recommend if to be included in bind, but you can try to assign this version.

File: app/Http/Controllers/ItemController.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+class ItemController extends Controller
+{
+
+   public function index($imobilId, Request $request, ModelLogger $logger) 
+     { 
+
+          $imobil= Imobil::findOrFail($imobilId);
+          
+         $logger->logModel2($user = $request->user(), $imobil);
```
Comment (egorbwork) [inline app/Http/Controllers/ItemController.php]: mistake

**Comment by egorbwork** [comment]
Still combination of 2 homeworks and final none of them is finished.

**Comment by liubamiron** [comment]
> Still combination of 2 homeworks and final none of them is finished.  

At 1 homework, create DebugRequestActivityLogger and ProductionRequestActivityLogger, request information. 
Where did I go wrong ?

**Comment by egorbwork** [comment]
Homework 3 -
image
Homework 4 -
image


I see changes for both homeworks here.

**Comment by liubamiron** [comment]
Sorry for 2 homeworks in one commit.

File: app/Http/Controllers/ItemController.php [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+class ItemController extends Controller
+{
+
+    public function index($imobilId, Request $request, ModelLogger $logger)
+    {
+
+        $imobil = Imobil::findOrFail($imobilId);
+
+        $logger->logModel2($user = $request->user(), $imobil);
```
Comment (egorbwork) [inline app/Http/Controllers/ItemController.php]: Not fixed.

File: app/Http/Middleware/LogActivityMiddleware.php [hunk 9]
```diff
@@ -2,29 +2,33 @@
...
+  class LogActivityMiddleware {
+
+    private RequestActivityLoggerInterface $logger;
+ 
+    
+    public function __construct(RequestActivityLoggerInterface $logger)
+   {
+       $this->logger = $logger;
+   }
+   // param Request $request
```
Comment (egorbwork) [inline app/Http/Middleware/LogActivityMiddleware.php]: Commented code won't be accepted.

**Review by egorbwork — CHANGES_REQUESTED** [review]
Still not fixed.

File: app/Http/Providers/AppServiceProvider.php (hunk 6 above)
Comment (liubamiron) [reply app/Http/Providers/AppServiceProvider.php]: include if in bind 87315ac

File: app/Http/Middleware/LogActivityMiddleware.php (hunk 9 above)
Comment (liubamiron) [reply app/Http/Middleware/LogActivityMiddleware.php]: delete commented code 9af722f

File: app/Http/Providers/AppServiceProvider.php (hunk 5 above)
Comment (liubamiron) [reply app/Http/Providers/AppServiceProvider.php]: removed definition

File: app/Services/ProductionRequestActivityLogger.php (hunk 2 above)
Comment (liubamiron) [reply app/Services/ProductionRequestActivityLogger.php]: what do you recommend, to collect in Production logger, (url, ip ..) ?

File: views/imobils/test.blade.php (hunk 3 above)
Comment (liubamiron) [reply views/imobils/test.blade.php]: fixed  6704343

File: app/Http/Controllers/ItemController.php (hunk 7 above)
Comment (liubamiron) [reply app/Http/Controllers/ItemController.php]: sorry, but i don't understand where is the mistake, when access imbilId,  send in log information( product id, price, and type of imobil(appartment or house)). in laravelLog I see this information ([2022-01-27 17:22:05] local.INFO: Uknown User accesed imobil with  {"id":2,"type":"APPARTMENT","price":300.0}

File: app/Services/DebugRequestActivityLogger.php (hunk 4 above)
Comment (liubamiron) [reply app/Services/DebugRequestActivityLogger.php]: collected data 3b51455

File: app/Http/Controllers/ItemController.php (hunk 8 above)
Comment (liubamiron) [reply app/Http/Controllers/ItemController.php]: change from logModel2 in logModelItem 7d4ba19

File: app/Services/ProductionRequestActivityLogger.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    protected function collectRequestData(Request $request): array
+    {
+        $name = $request->name;
+        $email = $request->email;
+        $department = $request->department;
+        $urlWithQueryString = $request->fullUrl();
+
+
+        return [
+            'url' . $urlWithQueryString,
```
Comment (egorbwork) [inline app/Services/ProductionRequestActivityLogger.php]: This one should work as associative array and should avoid usage of the data from request.

File: app/Services/DebugRequestActivityLogger.php [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+use Illuminate\Http\Request;
+use App\Models\LoggableInterface;
+use App\Models\Imobil;
+
+class DebugRequestActivityLogger extends AbstractRequestActivityLogger
+{
+    protected function collectRequestData(Request $request): array
+    {
+        return (array)$request->all(
```
Comment (egorbwork) [inline app/Services/DebugRequestActivityLogger.php]: request all could be a single key from debug log, and clearly without the provided keys.

Comment (egorbwork) [reply app/Services/DebugRequestActivityLogger.php]: and there should be present other data.

File: app/Console/Commands/Game.php [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+     * @var string
+     */
+    protected $signature = 'game:test';
+
+    /**
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'Command description';
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: please use adequate description.

File: app/Console/Commands/Game.php [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+use Illuminate\Contracts\Config\Repository as ConfigRepository;
+
+class Game extends Command
+{
+    /**
+     * The name and signature of the console command.
+     *
+     * @var string
+     */
+    protected $signature = 'game:test';
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: why test in command name?

File: app/Console/Commands/Game.php [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+    public function handle()
+    {
+
+        $lastVisit = $this->cacheRepository->get('lastCommandVisit', 'never');
+        $diedStats = $this->cacheRepository->get('diedCount', 0);
+        $surviveStats = $this->cacheRepository->get('surviveCount', 0);
+
+
+        $this->info("Command called last time: {$lastVisit}");
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: Single command for statistics and game, won't be accepted.

File: app/Console/Commands/Game.php [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        $this->info("Died Statistic: {$diedStats}");
+        $this->info("Survive Statistic: {$surviveStats}");
+
+        $this->cacheRepository->set('lastCommandVisit', date(\DateTime::ATOM), 86400);
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: should not be used here.

File: app/Console/Commands/Game.php [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+
+        $executionCicle = True;
+        $count = 1;
+        $table = [];
+        $answer = 'survive';
+        $res = 0;
+
+        while ($executionCicle) {
+            $random = $random = random_int(0, 2);
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: it should have random from 6 to emulate the revolver, and for statistics, I expect to receive the statistics, which number how many times was randomized.

**Review by egorbwork — CHANGES_REQUESTED** [review]
Still enough of things to fix.

File: app/Services/ProductionRequestActivityLogger.php (hunk 10 above)
Comment (liubamiron) [reply app/Services/ProductionRequestActivityLogger.php]: now it is associative array

File: app/Console/Commands/Game.php (hunk 12 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: now it is 'Russian roulette game'

File: app/Console/Commands/Game.php (hunk 13 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: delete test

File: app/Console/Commands/Game.php (hunk 15 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: change in 'visit game last time'

File: app/Console/Commands/Game.php (hunk 16 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: Change in random_int(1, 6), and add statistics for random numbers. a61787b

File: app/Console/Commands/Game.php [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+use Illuminate\Contracts\Config\Repository as ConfigRepository;
+
+class Game extends Command
+{
+    /**
+     * The name and signature of the console command.
+     *
+     * @var string
+     */
+    protected $signature = 'game1';
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: why `game1` and not simple `game`?

File: app/Console/Commands/Game.php [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+
+    private CacheRepository $cacheRepository;
+
+
+    /**
+     * Create a new command instance.
+     *
+     * @return void
+     */
+    public function __construct(CacheRepository $cacheRepository, ConfigRepository $configRepository)
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: ConfigRepository is still not needed here.

File: app/Console/Commands/Game.php [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+     * Execute the console command.
+     *
+     * @return int
+     */
+
+
+    public function handle()
+    {
+
+        $lastVisit = $this->cacheRepository->get('lastGameVisit', 'never');
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: should not be present

File: app/Console/Commands/Game.php [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+            $this->cacheRepository->set('Bad-numbers', $bad_numbers, 86400);
+    
+
+           
+            $res = $this->ask("Pull the triger!");
+
+            if ($random === 1) {
+                $answer = 'die';
+                $table[] = [$answer, $count, $random];
+                $this->table(['answer', 'attempts', 'random_nr'], $table);
```
Comment (egorbwork) [inline app/Console/Commands/Game.php:88]: Representation of statistics, was not changed.

File: app/Console/Commands/Game.php [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        $executionCicle = True;
+        $count = 1;
+        $table = [];
+        $table2 = [];
+        $answer = 'survive';
+        $res = 0;
+
+        while ($executionCicle) {
+
+            $random = $random = random_int(1, 6);
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: Why here is present twice assignment?

File: app/Console/Commands/Game.php [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+            $bad_numbers[$random] = $bad_numbers[$random] ?? 0;
+            $bad_numbers[$random]++;
+    
+            $this->cacheRepository->set('Bad-numbers', $bad_numbers, 86400);
+    
+
+           
+            $res = $this->ask("Pull the triger!");
+
+            if ($random === 1) {
```
Comment (egorbwork) [inline app/Console/Commands/Game.php]: The second number should be also randomized and this value should be kept in the statistics, in other way just 1 is a bad number.

**Review by egorbwork — CHANGES_REQUESTED** [review]
The commands was not split so it can't be accepted.

File: app/Console/Commands/Game.php (hunk 15 above)
Comment (egorbwork) [reply app/Console/Commands/Game.php]: The problem that is logic from example, you simply do not need it.

File: app/Console/Commands/Game.php (hunk 20 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php:88]: add bullet in chamber, and show when player was killed

File: app/Console/Commands/Game.php (hunk 17 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: change in 'game'

File: app/Console/Commands/Game.php (hunk 18 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: remove ConfigRepository

File: app/Console/Commands/Game.php (hunk 19 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: remove

File: app/Console/Commands/Game.php (hunk 21 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: delete one

File: app/Console/Commands/Game.php (hunk 22 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php]: now it is $random2 = random_int(1, 6) ebf4f7a

File: app/Console/Commands/Game.php (hunk 20 above)
Comment (egorbwork) [reply app/Console/Commands/Game.php:88]: still representing result in same command.

**Review by egorbwork — CHANGES_REQUESTED** [review]
Still no split in 2 commands.

File: app/Console/Commands/Game.php (hunk 20 above)
Comment (liubamiron) [reply app/Console/Commands/Game.php:88]: in table2 are other results, which show how many times was received bad number, and kill person $this->table(['bad nr.', 'dad times'], $table2);
